src/WebJobs.Script.WebHost/Security/SecretManager.cs

```diff
@@ -138,14 +138,78 @@ public virtual HostSecretsInfo GetHostSecrets()
                         }
                     };
 
-                    string secretsContent = ScriptSecretSerializer.SerializeFunctionSecrets(secrets);
-                    File.WriteAllText(secretFilePath, secretsContent);
+                    PersistFunctionSecrets(secrets, secretsFilePath);
+                }
+
+                // Read all secrets, which will run the keys through the appropriate readers
+                var keys = secrets.Keys.Select(k => ReadSecretValue(k)).ToList();
+
+                if (keys.Any(k => k.PersistenceStateStale))
+                {
+                    UpdateFunctionSecretsPersistenceState(keys, secretsFilePath);
                 }
 
-                return secrets.Keys.ToDictionary(s => s.Name, s => ReadSecretValue(s));
+                return secrets.Keys.ToDictionary(s => s.Name, s => s.Value);
             });
         }
 
+        private HostSecrets GenerateHostSecrets(string secretsFilePath)
+        {
+            // initialize with new secrets and save it
+            var hostSecrets = new HostSecrets
```

Comment:
It'd be good to get all this logic under some unit tests. E.g. if this entire method were internal so we can test it.